sql/src/main/java/io/crate/planner/node/ddl/UpdateSettingsPlan.java

```diff
@@ -80,32 +90,79 @@ public void executeOrFail(DependencyCarrier dependencies,
-            .transientSettings(buildSettingsFrom(transientSettings, params));
+            .persistentSettings(buildSettingsFrom(persistentSettings, eval, global))
+            .transientSettings(buildSettingsFrom(transientSettings, eval, global));
         OneRowActionListener<ClusterUpdateSettingsResponse> actionListener = new OneRowActionListener<>(
             consumer,
             r -> r.isAcknowledged() ? new Row1(1L) : new Row1(0L));
         dependencies.transportActionProvider().transportClusterUpdateSettingsAction().execute(request, actionListener);
     }
 
-    static Settings buildSettingsFrom(Map<String, List<Expression>> settingsMap, Row parameters) {
+    @VisibleForTesting
+    static Settings buildSettingsFrom(Map<Symbol, List<Symbol>> settingsMap,
+                                      Function<? super Symbol, Object> eval,
+                                      boolean global) {
         Settings.Builder settings = Settings.builder();
-        for (Map.Entry<String, List<Expression>> entry : settingsMap.entrySet()) {
-            String settingsName = entry.getKey();
+        for (Map.Entry<Symbol, List<Symbol>> entry : settingsMap.entrySet()) {
+            String settingsName = eval.apply(entry.getKey()).toString();
+
             if (CrateSettings.isValidSetting(settingsName) == false) {
                 throw new IllegalArgumentException("Setting '" + settingsName + "' is not supported");
             }
+            List<Symbol> value = entry.getValue();
 
-            List<Expression> value = entry.getValue();
-            // value can be null to reset settings
-            if (value == null) {
-                settings.put(settingsName, (String) null);
+            if (global) {
+                for (Symbol symbol : value) {
+                    for (String settingName : symbol.accept(SymbolToSettingVisitor.INSTANCE, settingsName)) {
+                        CrateSettings.checkIfRuntimeSetting(settingName);
+                    }
+                }
             } else {
-                CrateSettings.flattenSettings(settings, settingsName,
-                    ExpressionToObjectVisitor.convert(Iterables.getOnlyElement(value), parameters));
+                List<String> nameParts = CrateSettings.settingNamesByPrefix(settingsName);
+                if (nameParts.size() != 0) {
+                    throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                                                                     "GLOBAL Cluster setting '%s' cannot be used with SET SESSION / LOCAL",
+                                                                     settingsName));
+                }
             }
+            CrateSettings.flattenSettings(settings, settingsName, eval.apply(Iterables.getOnlyElement(value)));
         }
         return settings.build();
     }
+
+    private static class SymbolToSettingVisitor extends SymbolVisitor<String, Collection<String>> {
```

Comment:
can't this be replaced by using ``SymbolEvaluator`` early on the ``settingsMap``?

Reply:
The purpose of this `SymbolVisitor` is to concat on the key name of a nested object. e.g. https://github.com/crate/crate/blob/ed09e9dd61522bd0152701e1c8fad929783c6770/sql/src/test/java/io/crate/planner/node/ddl/UpdateSettingsPlanTest.java#L95
Is this also possible using the `SymbolEvaluator` ?

Reply:
Conceptually I'd prefer to map all Symbols early to Objects. I think `CrateSettings.flattenSettings` could be used afterwards to get a list of flat settings.

Reply:
Ok, thanks. understood.